### ISSUE TITLE
StructuredText Types can't tell links and blocks apart

### ISSUE THREAD

**Issue by thebuilder** [datocms/react-datocms#33]
Trying to build a StructuredText block that contains both `blocks` and `links` - The output of this works fine, but the actual types are throwing errors.

```graphql
article {
      value
      links {
        ...ArticlePageLink
        ...ExternalUrlLink
        ...ContentPageLink
        ...ContactPageLink
        ...HomePageLink
      }
      blocks {
        ... on ExternalVideoRecord {
          __typename
          id
          videoLink {
            provider
            url
          }
        }
        ... on VideoRecord {
          __typename
          id
          video {
            video {
              mp4Url
              thumbnailUrl(format: jpg)
              streamingUrl
            }
          }
        }
        ... on ImageRecord {
          __typename
          id
          image {
            url
            width
            height
            title
            alt
          }
        }
      }
    }
```

This throws a huge error on the `data` field.

```
TS2322: Type '{ __typename: "ArticlePageModelArticleField"; value: any; links: (({ __typename: "ArticlePageRecord"; } & ArticlePageLinkFragment) | ({ __typename: "ContactPageRecord"; } & ContactPageLinkFragment) | ({ ...; } & ContentPageLinkFragment) | ({ ...; } & ExternalUrlLinkFragment) | ({ ...; } & HomePageLinkFragment))[]; b...' is not assignable to type 'Document | Node | StructuredText<{ __typename: "ExternalVideoRecord"; id: any; videoLink?: { __typename: "VideoField"; provider?: string | undefined; url?: string | undefined; } | undefined; } | { ...; } | { ...; }> | null | undefined'.   Type '{ __typename: "ArticlePageModelArticleField"; value: any; links: (({ __typename: "ArticlePageRecord"; } & ArticlePageLinkFragment) | ({ __typename: "ContactPageRecord"; } & ContactPageLinkFragment) | ({ ...; } & ContentPageLinkFragment) | ({ ...; } & ExternalUrlLinkFragment) | ({ ...; } & HomePageLinkFragment))[]; b...' is not assignable to type 'Document | Node | StructuredText<{ __typename: "ExternalVideoRecord"; id: any; videoLink?: { __typename: "VideoField"; provider?: string | undefined; url?: string | undefined; } | undefined; } | { ...; } | { ...; }> | null | undefined'.     Type '{ __typename: "ArticlePageModelArticleField"; value: any; links: (({ __typename: "ArticlePageRecord"; } & ArticlePageLinkFragment) | ({ __typename: "ContactPageRecord"; } & ContactPageLinkFragment) | ({ ...; } & ContentPageLinkFragment) | ({ ...; } & ExternalUrlLinkFragment) | ({ ...; } & HomePageLinkFragment))[]; b...' is not assignable to type 'StructuredText<{ __typename: "ExternalVideoRecord"; id: any; videoLink?: { __typename: "VideoField"; provider?: string | undefined; url?: string | undefined; } | undefined; } | { __typename: "ImageRecord"; id: any; image?: { ...; } | undefined; } | { ...; }>'.       Types of property 'links' are incompatible.         Type '(({ __typename: "ArticlePageRecord"; } & ArticlePageLinkFragment) | ({ __typename: "ContactPageRecord"; } & ContactPageLinkFragment) | ({ ...; } & ContentPageLinkFragment) | ({ ...; } & ExternalUrlLinkFragment) | ({ ...; } & HomePageLinkFragment))[]' is not assignable to type '({ __typename: "ExternalVideoRecord"; id: any; videoLink?: { __typename: "VideoField"; provider?: string | undefined; url?: string | undefined; } | undefined; } | { __typename: "ImageRecord"; id: any; image?: { ...; } | undefined; } | { ...; })[]'.           Type '({ __typename: "ArticlePageRecord"; } & ArticlePageLinkFragment) | ({ __typename: "ContactPageRecord"; } & ContactPageLinkFragment) | ({ ...; } & ContentPageLinkFragment) | ({ ...; } & ExternalUrlLinkFragment) | ({ ...; } & HomePageLinkFragment)' is not assignable to type '{ __typename: "ExternalVideoRecord"; id: any; videoLink?: { __typename: "VideoField"; provider?: string | undefined; url?: string | undefined; } | undefined; } | { __typename: "ImageRecord"; id: any; image?: { ...; } | undefined; } | { ...; }'.             Type '{ __typename: "ArticlePageRecord"; } & ArticlePageLinkFragment' is not assignable to type '{ __typename: "ExternalVideoRecord"; id: any; videoLink?: { __typename: "VideoField"; provider?: string | undefined; url?: string | undefined; } | undefined; } | { __typename: "ImageRecord"; id: any; image?: { ...; } | undefined; } | { ...; }'.               Type '{ __typename: "ArticlePageRecord"; } & ArticlePageLinkFragment' is not assignable to type '{ __typename: "VideoRecord"; id: any; video?: { __typename: "FileField"; video?: { __typename: "UploadVideoField"; mp4Url?: string | undefined; thumbnailUrl: string; streamingUrl: string; } | undefined; } | undefined; }'.                 Types of property '__typename' are incompatible.                   Type '"ArticlePageRecord"' is not assignable to type '"VideoRecord"'.
```

TLDR, I think it boils down to the last part: 
```
Types of property '__typename' are incompatible.                  
 Type '"ArticlePageRecord"' is not assignable to type '"VideoRecord"'
```

The `renderInlineRecord`, `renderLinkToRecord` and `renderBlock` are checking for all possible records instead of just `links["__typename"]` or `blocks["__typename"]`, so if I assume the value must be one of my link types, it will also throw an error.

* If i remove the `links` from the GraphQL query, then it works fine.
* If i remove the `blocks` from the GraphQL query, then it also works fine.

**Expected output**
The `<StructuredText>` output should correctly infer the `__typename` of `links` and `blocks`, and use type guards to ensure the `render` methods only get the relevant types.

**Comment by stefanoverna** [comment]
Hi @thebuilder, please take a look at this TS test:

https://github.com/datocms/react-datocms/blob/master/src/StructuredText/__tests__/index.test.tsx#L129-L227

We're using both links and blocks, and every type seems to be working there, so I'm afraid I have to ask you to share a small project that reproduces the error you're mentioning.

Thanks!

**Comment by thebuilder** [comment]
I'll see what I can do.

**Comment by thebuilder** [comment]
I'm generating the queries with https://graphql-code-generator.com, so it could be related to how the types are generated there - But they are otherwise perfect.
